src/pycountry/db.py

```diff
@@ -1,32 +1,68 @@
 import json
 import logging
 import threading
+import warnings
+from typing import Any, Callable, Dict, Iterator, List, Optional, Type, TypeVar
 
 logger = logging.getLogger("pycountry.db")
 
 
 class Data:
-    def __init__(self, **fields):
+    def __init__(self, **fields: str):
         self._fields = fields
 
     def __getattr__(self, key):
-        if key not in self._fields:
-            raise AttributeError
-        return self._fields[key]
-
-    def __setattr__(self, key, value):
+        if self.__class__.__name__ == "Country":
```

Comment:
Checking for class names seems a bit hacky. The Country class could simply override the __getattr__ method IMHO.

Reply:
@ctheune I agree I was trying for a long time to try and get this to work. 

[This is the code](https://github.com/Jamie-/pycountry/commit/8d2341a288c0ea39cc4cd675173130f13a5382f2) that @Jamie- originally wrote that had a __getattr__ method defined within the country class

When I was testing this specific state it seemed to work. Since then I can't seem to get it to work as Jamie did. I wonder if there is some type of conflict between @Jamie- 's code and @eliasmistler which did change the Database._load method slightly 

Somebody who is more well versed in object oriented programming may see the issue easier than I.

Reply:
The code from @Jamie- looks fine, even with @eliasmistler changes on top from what I'd expect. I'd have to poke this more closely. Did you get a specific error or just wrong results?

Reply:
@ctheune if I try to implement like this 

``` python
class Data:
    def __init__(self, **fields: str):
        self._fields = fields

    def __getattr__(self, key):
        # For classes other than Country
        if key in self._fields:
            return self._fields[key]
        raise AttributeError()

...

class Country(Data):
    def __getattr__(self, key):
        print("USING COUNTRY SEARCH")
        if key in ("common_name", "official_name"):
            # First try to get the common_name or official_name
            value = self._fields.get(key)
            if value is not None:
                return value
            # Fall back to name if common_name or official_name is not found
            name = self._fields.get("name")
            if name is not None:
                warning_message = f"Country's {key} not found. Country name provided instead."
                warnings.warn(warning_message, UserWarning)
                return name
        # For other keys, delegate to the parent class's __getattr__
        return super().__getattr__(key)

```

I get this error

```
>>> import pycountry
>>> united_states = pycountry.countries.get(alpha_2='US')
>>> united_states.common_name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/nate.schimmoller/Documents/GitHub/pycountry/src/pycountry/db.py", line 18, in __getattr__
    raise AttributeError()
AttributeError
```

Given it never prints USING COUNTRY SEARCH I'm assuming that it is using the parent (Data) class' `__getattr__` method

Reply:
Ok, this is nasty. The interesting part is in the weird data class mangling. I've rewritten this in a separate branch (`cleanup-dataclasses-meta`) that you can merge onto this one (and clean up the history if you like). The diff should help enlighten what's going on. Short story: the Country class is never used as it always synthesizes a class that is equally named (and thus confusing).